lib/cfg.js

```diff
@@ -0,0 +1,368 @@
+    };
+
+    var code = [];
+    var first = 0;
+    var last = asmArr.length;
+    
+    if(first == last) return code;
+
+    
+    while(first != last-1){
+        if(isCode(asmArr, first))
+            code.push(JSON.parse(JSON.stringify(asmArr[first])));
+        
+        ++first;
+    }
+    if(asmArr[first].source){
+        code.push(JSON.parse(JSON.stringify(asmArr[first])));
+    }
+
+    return code;
+}
+
+    var gccX86 = {
+                filterData:  seperateCodeFromData,
+                isFunctionEnd:function(x)  { return ((x[0] != ' ') && (x[0] != '.') &&
+                                                             (x.indexOf(':') != -1)) ;},
+
+                isBasicBlockEnd: function(x) { return x[0] == ".";} ,
+
+                getInstructionType:function(inst) {
+                                            inst = inst.trim();
+                                            if (inst === "") return -1;
+                                            else if(inst.includes("jmp")) return 0;
+                                            else if(inst[0] === 'j') return 1;
+                                            else if(!inst.includes("ret")) return 2;
+                                            else return 3;
+                                          },
+
+                extractNodeName:function(inst) {
+                                     var name = inst.match(/.L\d+/);
+                                     return name + ":";
+                                  },
+
+                isJmpInstruction: function(x) { var trimed = x.trim(); return ( trimed[0] == 'j');}
+             };
+
+
+
+
+ControlFlowGraph.prototype.splitToFunctions = function (asmArr, isEnd) {
+    if(asmArr.length === 0)   return [];
+    var result = [];
+    var first = 0;
+    var last = asmArr.length;
+    var fnRange = {start:first, end:null};
+    ++first;
+    while(first != last) {
+        if(isEnd(asmArr[first].text)) {
+            fnRange.end = first;
+            result.push(JSON.parse(JSON.stringify(fnRange)));
+            fnRange.start = first;
+        }
+        ++first;
+    }
+
+    fnRange.end = last;
+    result.push(fnRange);//possiblitiy of bug?! because didn't clone(JSON.parse(JSON.stringify(...))
+    return result;
+};
+
+ControlFlowGraph.prototype.splitToBasicBlocks = function (asmArr,  range, isEnd, isJmp) {
+    var first = range.start;
+    var last = range.end;
+    if(first == last) return [];
+    ++first;
+
+    var rangeBb = {nameId:"start" , start:first , end:null, actionPos:[]};
+    var result = [];
+
+    var resetRangeWith = function(rangeBb,  nameId, start) {
+            rangeBb.nameId = nameId;//.L1:
+            rangeBb.start = start;//after .L1:
+            rangeBb.actionPos = [];
+    };
+
+    while(first != last) {
+        var inst = asmArr[first].text;
+        if(isEnd(inst)) {
+            rangeBb.end = first;
+            result.push(JSON.parse(JSON.stringify(rangeBb)));
```

Comment:
_.clone if needed, as above. Or just change `resetRangeWith` to return a new range each time. The "return a copy of mutated data" idiom makes it much easier to reason about what might change, rather than having a single object and keep modifying it. You're paying the price of making a copy each time anyway, either with `clone` or `JSON.parse(JSON.stringify)` or just something like:

```
var newRangeWith = function(oldRange, nameId, start) {
  return {nameId: nameId, start: start, actionPos: [], end: oldRange.end};
};
....
if (isEnd) {
  rangeBb.end = first;
  result.push(rangeBb);
  rangeBb = newRangeWith(rangeBb, inst, first);
}
```

Reply:
done